### ISSUE TITLE
MacOS + webrtc-sendrecv.py = Illegal instruction: 4

### ISSUE THREAD

**Issue by ykazakov** [centricular/gstwebrtc-demos#40]
I have some problems running the Python example on macOS: the connection starts, in the browser window (chrome and firefox) I see in the flying ball and hear noise, but after a few seconds the program crashes with "Illegal instruction: 4". A window with the video stream coming from the browser does not open. In contrast, if I run the compiled C example, I see the video from the browser. If I set in the browser {"video":false,"audio":true}, the Python program works fine, but the video is not transmitted.

I did some investigation, and I think that the crash is caused by the threading issues because on [MacOS the UI operations should be called from the main thread](http://gstreamer-devel.966125.n4.nabble.com/OpenGL-renderer-window-tp4686092p4686093.html). But in the Python program the pipeline is created on the asyncio event loop. Unfortunately I am completely newbie in Python so I do not see any easy fix here.

gsteamer 1.14.2 installed on MacOS 10.13.6 by homebrew with the following options:
```
brew install gstreamer
brew install gst-plugins-base --with-opus
brew install gst-plugins-good --with-speex --with-libvpx --with-jack
brew install gst-plugins-bad --with-libnice --with-opus --with-srtp
brew install gst-python
```
```
$ python3 --version
Python 3.7.0
```
The test pipeline works fine:
```
gst-launch-1.0 videotestsrc pattern=ball ! videoconvert ! autovideosink
```
Here is a sample run:
```
$ python3 webrtc-sendrecv.py 1721
Sending offer:
v=0
o=- 3303706454607302771 0 IN IP4 0.0.0.0
s=-
t=0 0
a=ice-options:trickle
a=msid-semantic:WMS sendrecv
m=video 9 UDP/TLS/RTP/SAVPF 97
c=IN IP4 0.0.0.0
a=setup:actpass
a=ice-ufrag:XBq4jDVORdgcPIcLgro04jxbO0Y5xi1c
a=ice-pwd:yjpMNpC9jlhlnvLo+4IPlMdRLUGuRDzH
a=sendrecv
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:97 VP8/90000
a=rtcp-fb:97 nack pli
a=framerate:30
a=ssrc:1949809321 msid:user2948245404@host-9bf5ae61 webrtctransceiver0
a=ssrc:1949809321 cname:user2948245404@host-9bf5ae61
a=mid:video0
a=fingerprint:sha-256 BB:CF:2D:0F:56:B6:EB:52:74:B9:13:E6:35:49:75:4D:13:C4:4A:F4:9C:07:82:02:F3:0A:DF:1D:D9:6A:F3:EA
m=audio 9 UDP/TLS/RTP/SAVPF 96
c=IN IP4 0.0.0.0
a=setup:actpass
a=ice-ufrag:VMg4ywOfav8gWkIMwaNBgVUJgUjoUeAl
a=ice-pwd:PxkPCSMFlgIKu856GIbqiMHnElhTHAJf
a=sendrecv
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:96 OPUS/48000/2
a=rtcp-fb:96 nack pli
a=fmtp:96 sprop-maxcapturerate=48000;sprop-stereo=0
a=ssrc:2488823075 msid:user2948245404@host-9bf5ae61 webrtctransceiver1
a=ssrc:2488823075 cname:user2948245404@host-9bf5ae61
a=mid:audio1
a=fingerprint:sha-256 BB:CF:2D:0F:56:B6:EB:52:74:B9:13:E6:35:49:75:4D:13:C4:4A:F4:9C:07:82:02:F3:0A:DF:1D:D9:6A:F3:EA

Received answer:
v=0
o=- 5883944315674643207 2 IN IP4 127.0.0.1
s=-
t=0 0
a=msid-semantic: WMS y8E09mXfXUFm2m2rgN7NgBE2D28tgJEM6TA3
m=video 9 UDP/TLS/RTP/SAVPF 97
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:TuJQ
a=ice-pwd:jvwwL3myUqdhCz8e7ifGdJfY
a=ice-options:trickle
a=fingerprint:sha-256 14:98:38:33:92:92:82:1E:69:C8:93:D9:B1:4D:60:B1:09:7C:60:C5:F4:44:58:61:09:44:90:3B:6F:B7:BA:3D
a=setup:active
a=mid:video0
a=sendrecv
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:97 VP8/90000
a=rtcp-fb:97 nack pli
a=ssrc:714727404 cname:qZWzu+tTH9Ykc9TN
a=ssrc:714727404 msid:y8E09mXfXUFm2m2rgN7NgBE2D28tgJEM6TA3 32c1ea82-212e-4214-9a01-823dd6c02e0d
a=ssrc:714727404 mslabel:y8E09mXfXUFm2m2rgN7NgBE2D28tgJEM6TA3
a=ssrc:714727404 label:32c1ea82-212e-4214-9a01-823dd6c02e0d
m=audio 9 UDP/TLS/RTP/SAVPF 96
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:YKTI
a=ice-pwd:oHJ1ptMvPMigaRwZjxMHMkD/
a=ice-options:trickle
a=fingerprint:sha-256 14:98:38:33:92:92:82:1E:69:C8:93:D9:B1:4D:60:B1:09:7C:60:C5:F4:44:58:61:09:44:90:3B:6F:B7:BA:3D
a=setup:active
a=mid:audio1
a=sendrecv
a=rtcp-mux
a=rtpmap:96 OPUS/48000/2
a=fmtp:96 minptime=10;useinbandfec=1
a=ssrc:2681704157 cname:qZWzu+tTH9Ykc9TN
a=ssrc:2681704157 msid:y8E09mXfXUFm2m2rgN7NgBE2D28tgJEM6TA3 6c009868-51be-4c79-b76b-3ed1bfb9eb68
a=ssrc:2681704157 mslabel:y8E09mXfXUFm2m2rgN7NgBE2D28tgJEM6TA3
a=ssrc:2681704157 label:6c009868-51be-4c79-b76b-3ed1bfb9eb68

Illegal instruction: 4
```

**Comment by ystreet** [comment]
As you say, MacOS requires that UI operations occur on the main thread. Unfortunately, that's a hard thing to do when you don't control the thread loop.  The display element in question (glimagesink) does contain code for attempting to initialize everything correctly when using GMainLoop and the GStreamer binaries for MacOS also contain a patch that iterate the main CFRunLoop when iterating the main GMainLoop.  Without that one needs to run the CFRunLoop manually like in: https://github.com/sdroege/gstreamer-rs/blob/master/examples/src/examples-common.rs

In short, for MacOS, the python example would need to be written around either GMainLoop instead of asyncio or attempt to run the main CFRunLoop from python.  Can you try the c or rust versions instead?

**Comment by ykazakov** [comment]
As I mentioned, the C code works fine. The problem appears only with the Python code. Thanks for the pointer to a workaround!
